File: src/lan_java/prbm_greed/MakeBigNum.java
package lan_java.prbm_greed;

import java.util.Stack;

public class MakeBigNum {

	public static void main(String[] args) {
		String number = "999";
		int k = 2;
		System.out.println(solution2(number, k));		
	}
	
    public static String solution(String number, int k) {
    	StringBuilder sb = new StringBuilder();

    	int numLen = number.length();
    	int startIndex=0;
    	
    	for(int i=k; i<numLen; i++) {
    		char tempMax=number.charAt(startIndex);
    		int tempMaxIndex=startIndex;
    		for(int j=startIndex; j<=i; j++) {
    			if(number.charAt(j)>tempMax) {
    				tempMax=number.charAt(j);
    				tempMaxIndex=j;
    				
    			}
    		}
    		sb.append(number.charAt(tempMaxIndex));
    		startIndex=tempMaxIndex+1;
    	}
    	return sb.toString();  	    	
    }
    
    public static String solution2(String number, int k) {
        char[] result = new char[number.length() - k];
        Stack<Character> stack = new Stack<>();

        for (int i=0; i<number.length(); i++) {
            char c = number.charAt(i);
            while (!stack.isEmpty() && stack.peek() < c && k-- > 0) {
                stack.pop();
            }
            stack.push(c);
        }
        
        for (int i=0; i<result.length; i++) {
            result[i] = stack.get(i);
        }
        return new String(result);
    }
}

File: src/lan_python/prbm_greed/MakeBigNum.py
def solution(number, k):
    stack = [number[0]]
    for num in number[1:]:
        while len(stack) > 0 and stack[-1] < num and k > 0:
            k -= 1
            stack.pop()
        stack.append(num)
    if k != 0:
        stack = stack[:-k]
    return ''.join(stack)


## test case 10 시간초과 발생 
def solution2(number, k):
    result = []
    numLen = len(number)
    startIndex = 0

    for i in range(k, numLen):
        tempMaxIndex = startIndex
        for j in range(startIndex, i+1):
            if number[j]>number[tempMaxIndex]:
                tempMaxIndex = j
        result.append(number[tempMaxIndex])
        startIndex = tempMaxIndex+1
    
    return ''.join(result)

a = solution2('999', 2)
print(a)